File: src/1235-recommend/lab02.ts
import * as alcoholMap from '../../docs/alcohol-map.json';
import {LeafData, ExtendedLeafData} from './types/data';
import {TreeNode} from './types/node';
import {getLeavesData, normalizeLeavesData} from './algorithms/leaves';
import {calculateEuclideanDistance} from './metrics/euclidean';
import {calculateManhattanDistance} from './metrics/manhattan';
import {calculateTreeDistance} from './metrics/tree';
import {calculatePearsonCorrelationCoefficient} from './metrics/pearson';
import {calculateCustomDistance} from './metrics/custom';
import {readProduct} from './io/readProduct';


const lab02 = () => {
	const root: TreeNode = alcoholMap;
	const leavesData: LeafData[] = getLeavesData(root);
	const extendedLeavesData: ExtendedLeafData[] = normalizeLeavesData(leavesData);
	const products: string[] = leavesData.map((leafData: LeafData): string => leafData.name);

	console.log('List of products: ', products);

	const N = 10;
	for (let i = 1; i < N; ++i) {
		console.log(`\nITERATION ${i}`);
		const product1 = readProduct('first', products);
		const product2 = readProduct('second', products);

		const extendedLeafData1 = extendedLeavesData[products.indexOf(product1)];
		const extendedLeafData2 = extendedLeavesData[products.indexOf(product2)];

		const euclideanDistance = calculateEuclideanDistance(extendedLeafData1, extendedLeafData2);
		const manhattanDistance = calculateManhattanDistance(extendedLeafData1, extendedLeafData2);
		const treeDistance = calculateTreeDistance(root, extendedLeafData1, extendedLeafData2);
		const pearsonCorrelationCoefficient = calculatePearsonCorrelationCoefficient(extendedLeafData1, extendedLeafData2);
		const customDistance = calculateCustomDistance(root, extendedLeafData1, extendedLeafData2);

		console.log(`Euclidean distance: ${euclideanDistance}`);
		console.log(`Manhattan distance: ${manhattanDistance}`);
		console.log(`Tree distance: ${treeDistance}`);
		console.log(`Pearson correlation coefficient: ${pearsonCorrelationCoefficient}`);
		console.log(`Custom distance: ${customDistance}`);
	}
};

lab02();

File: src/678-va-dialog/lab02.py
from typing import List

from algorithms.leaves import get_leaves_data, normalize_leaves_data
from io_util.read_json import read_map
from io_util.read_product import read_product
from metrics.custom import calculate_custom_distance
from metrics.euclidean import calculate_euclidean_distance
from metrics.manhattan import calculate_manhattan_distance
from metrics.tree import calculate_tree_distance
from metrics.pearson import calculate_pearson_correlation_coefficient
from ttypes.data import LeafData, ExtendedLeafData
from ttypes.node import TreeNode


def main() -> None:
    root: TreeNode = read_map()
    leaves_data: List[LeafData] = get_leaves_data(root)
    extended_leaves_data: List[ExtendedLeafData] = normalize_leaves_data(leaves_data)
    products: List[str] = [leaf_data['name'] for leaf_data in leaves_data]

    print('List of products: ', products)

    n = 10
    for i in range(n):
        print(f'\nITERATION {i}')
        product1 = read_product('first', products)
        product2 = read_product('second', products)

        e1 = extended_leaves_data[products.index(product1)]
        e2 = extended_leaves_data[products.index(product2)]

        euclidean_distance = calculate_euclidean_distance(e1, e2)
        manhattan_distance = calculate_manhattan_distance(e1, e2)
        tree_distance = calculate_tree_distance(root, e1, e2)
        pearson_correlation_coefficient = calculate_pearson_correlation_coefficient(e1, e2)
        custom_distance = calculate_custom_distance(root, e1, e2)

        print(f'Euclidean distance: {euclidean_distance}')
        print(f'Manhattan distance: {manhattan_distance}')
        print(f'Tree distance: {tree_distance}')
        print(f'Pearson correlation coefficient: {pearson_correlation_coefficient}')
        print(f'Custom distance: {custom_distance}')


if __name__ == '__main__':
    main()
